### ISSUE TITLE
Some of mlpack tests are failing on aarch64 machines

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#2869]
Hello everyone,

Following the work on this pull request #2531, we have noticed recently that there are several mlpack tests that are failing when executed on aarch64 architecture (arm architecture 64 bits). These tests are related to:

* FastMKSTest: Failing
* AKFNTest: Failing
* AKNNTest: Failing
* KFNTest: Failing
* KNNTest: Failing
* KMeansTest: Failing (DTNNCoverTreeTest) precisely
* KDETest: Failing (StandardCoverTreeGaussianTest)
* RangeSearchTest: Failing (RSCoverTreeTest)
* EmstTest: Failing (EmstCoverTreeTest)
* KMeansMainTest: Failing (AlgorithmSimilarTest)
* KDEMainTest: Failing (KDEMonoResultMain)

Note that, not all the tests inside these tags are failing, but some of them (Those related to dual cover tree). After running these failing tests, you should expect to enter an infinite loop that never ends. Finally, you will have to use Ctrl-C to stop it.

Usually, we cross-compile binaries from x86_64 bit machines to aarch64 machines, however lately I have found that 
the same issue is produced even though if I compiled these binaries locally on aarch64 (depending on which test you are executing)

I have tested several tests, the following tests (I thought they were relevant) are passing without any issue:
* TreeTest: Passing
* RectangleTreeTraitsTest: Passing
* KernelTest: Passing
* CosineTreeTest: Passing
* KRANNTest: Passing
* HoeffdingTreeTest: Passing
* SerializationTest: Passing
* TreeTraitsTest: Passing
* UBTreeTest: Passing
* VanatagePointTreeTest: Passing

If you have any test to execute that you think might be relevant to this issue, please let me know, and I will add the results back here. To reproduce the issue you need to have access to aarch64 machine through qemu emulator or physical hardware, if you have an idea and you would like to try, but you do not have access to the machine, please leave in a comment and a piece of code to add, modify or to examine and I will be very happy to the tests.

#### Here are some details about the test environment
 * Raspberry Pi 3 with ubuntu 20. (Zin64 on Jenkins)
 * Viking machine on Jenkins
 * version of mlpack: The master version
 * operating system: arch Linux, ubuntu
 * cross-compiler: the mainstream gcc 10 cases more of this issue, I am using this version of buildroot [https://toolchains.bootlin.com/](cross-compilers)
 * compiler: the mainstream version of gcc that you can find in any aarch64

#### Interesting info:
Try to use the buildroot cross-compiler set, since it is the only one that produced binaries that were having the same failure as the local compiler for an aarch64. However, if you use the main gcc cross-compiler. I think it should be fine.  

#### Steps to reproduce
Cross-compile mlpack tests for aarch64 or locally build these tests from aarch64 machine.
`./bin/mlpack_test  [FastMKSTest]`

PS: I will update the issue regularly with any tests that are passing that are using the cover tree

**Comment by rcurtin** [comment]
I tried to reproduce this today on my Samsung Chromebook Plus, but I'm actually not even able to compile the tests.  I don't have enough RAM. :)  Is there any chance you can provide some more details on how to cross-compile with buildroot?  I'm not experienced with cross-compilation, so if you can provide a quick guide so I can reproduce the issue, then I can dig in and figure out what's wrong. :+1:

**Comment by shrit** [comment]
Of course :smile: 
1) First you need to get the last stable version of buildroot: https://toolchains.bootlin.com/
2) You need to untar the toolchain into you home dir.
3) You need to get my mlpack branch for aarch64.
4) You have to set two variables the toolchain path and the sysroot either from the command line in cmake or directly in the file to have something similar to: 
`TOOLCHAIN_PATH=/home/ryan/aarch64--glibc--stable-2020.08-1/bin`
`SYSROOT_PATH=home/ryan/aarch64--glibc--stable-2020.08-1/aarc64-buildroot-linux-gnu/sysroot` 
Finally after setting these variable, start Cmake with the following command:
`cmake -DDEBUG=OFF -DPROFILE=OFF -DBUILD_MARKDOWN_BINDINGS=OFF -DBUILD_PYTHON_BINDINGS=OFF -DBUILD_GO_BINDINGS=OFF -DBUILD_JULIA_BINDINGS=OFF -DBUILD_R_BINDINGS=OFF -DCROSS_COMPILE=ON -DAUTO_DOWNLOAD_DEPS=ON -DCMAKE_TOOLCHAIN_FILE=../board/arm-64bit-toolchain.cmake -DBOARD_NAME="rpi3" -DGFORTRAN_LIBRARY=/home/ryan/aarch64--glibc--stable-2020.08-1/aarch64-buildroot-linux-gnu/lib64/libgfortran.a ../`

**Comment by rcurtin** [comment]
I tried following the directions; thanks for providing those.  I had to do a little bit of adaptation (mostly changing `TOOLCHAIN_PREFIX`), but nothing too serious.  I also had to download the dependencies and install them into the right place, but it seems like maybe something has been linked wrong?  I get this when I build:

```
...
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_lock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:749: undefined reference to `pthread_mutex_lock'
/home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: /home/ryan/src/aarch64--glibc--stable-2020.08-1/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o): in function `__gthread_mutex_unlock':
/opt/output/build/host-gcc-final-9.3.0/build/aarch64-buildroot-linux-gnu/libgcc/./gthr-default.h:779: undefined reference to `pthread_mutex_unlock'
collect2: error: ld returned 1 exit status
make[2]: *** [src/mlpack/methods/cf/CMakeFiles/mlpack_cf.dir/build.make:106: bin/mlpack_cf] Error 1
make[1]: *** [CMakeFiles/Makefile2:2810: src/mlpack/methods/cf/CMakeFiles/mlpack_cf.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

**Comment by shrit** [comment]
@rcurtin it seems that it is not able to find the `libpthread.a`, the path for libpthread.a is the same for libgfortran.a `path_to/sysroot/usr/lib`. Would it be possible to define the path manually for libpthread?
I will figure out the issue later, it should be able to define that without any issue.

**Comment by rcurtin** [comment]
Ok---I've managed to reproduce things.  For `pthread`, I just needed to add `pthread` to the list of libraries to link against (e.g. `COMPILER_SUPPORT_LIBRARIES` in CMake).  We'll probably need to find some better way to do that, but for now this works.  (I'll post more about that in #2531.)

Anyway, on my Chromebook Plus, I don't get infinite loops, but I do get test failures, which are probably the same thing:

```
[ryan@smash mlpack-test]$ ./mlpack_test "[RangeSearchTest]"
mlpack version: mlpack git-6967cf447
armadillo version: 10.3.0 (Sunrise Chaos)
Filters: [RangeSearchTest]

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mlpack_test is a Catch v2.13.4 host application.
Run with -? for options

-------------------------------------------------------------------------------
TrainTreeTest
-------------------------------------------------------------------------------
/home/ryan/src/mlpack/src/mlpack/tests/range_search_test.cpp:1140
...............................................................................

/home/ryan/src/mlpack/src/mlpack/tests/range_search_test.cpp:1140: FAILED:
due to unexpected exception with message:
  std::bad_alloc

-------------------------------------------------------------------------------
RSModelTest
-------------------------------------------------------------------------------
/home/ryan/src/mlpack/src/mlpack/tests/range_search_test.cpp:1277
...............................................................................

/home/ryan/src/mlpack/src/mlpack/tests/range_search_test.cpp:1277: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  std::bad_alloc

-------------------------------------------------------------------------------
RSModelMonochromaticTest
-------------------------------------------------------------------------------
/home/ryan/src/mlpack/src/mlpack/tests/range_search_test.cpp:1364
...............................................................................

/home/ryan/src/mlpack/src/mlpack/tests/range_search_test.cpp:1364: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  std::bad_alloc

===============================================================================
test cases:       22 |       19 passed | 3 failed
assertions: 13332696 | 13332693 passed | 3 failed
```

This should give me enough to dig into, but it may be a little while until I have more information (this is a tedious thing to dig into :)).

**Comment by shrit** [comment]
Thanks for the feedback, Yes, it is exactly the same problem, It used to gives `std::bad_alloc` when it gets out of the infinite loops eventually (in rare cases).
I will take care of the `pthread`

**Comment by shrit** [comment]
@rcurtin I have spent time trying to debug the `pthread` issue. I have found that when `find_package(thread)` is called in CMakeListe it is only finding the header `pthread.h` and not the library itself. Therefore, making mlpack dependant on the header rather than the library,  on the `x86_64` system this passes fine, because the linker script knows where to look for `pthread`, only `-lpthread` is required. However, this is not the case when cross-compiling

Therefore, when cross-compiling, it might be a good idea to specify where is the `pthread` library is located the same way as we are doing now `gfortran`

To resolve this issue I have added `PTHREAD_LIBRARY` variable, so user can add it to the command line, it might seem like a good solution since each distro/toolchain has its own path for the `pthread` and `gfortran`, `Ubuntu != Arch != buildroot`, at least in buildroot they have the same path.

Let me know what do you think about this solution.

**Comment by rcurtin** [comment]
I think that this will fix the issue:

https://gist.github.com/85e4f97332ae58ea7f4e9aa5036b9080

Basically, it seems that using reverse iterators with a `std::map<>` on cross-compiled arm64 code causes failures.  Maybe this is a compiler bug, but I did not dig deeper (or try a different compiler).  Instead, I simply worked around the issue by adapting the cover tree to use forward-mode iterators (which honestly is clearer code anyway).

I didn't open a PR, because this was done against the `supported-platforms` branch, so probably it's best to just apply that as part of #2531. :+1:
